File: Renaissance-ProgrammingPathshala/M2/Mathematics/HomeWork_3/ColouredField.cpp
#include <bits/stdc++.h>
using namespace std;

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);
    int n, m, k, t;
    cin >> n >> m >> k >> t;
    vector<string> crops{"Carrots", "Kiwis", "Grapes"};
    vector<vector<int>> wasteCells(k, vector<int>(2));
    map<pair<int, int>, bool> isWasteCell;
    for (int i = 0; i < k; i++)
    {
        cin >> wasteCells[i][0] >> wasteCells[i][1];
        isWasteCell[{wasteCells[i][0], wasteCells[i][1]}] = true;
    }
    while (t--)
    {
        int x, y;
        cin >> x >> y;
        if (isWasteCell[{x, y}])
        {
            cout << "Waste";
        }
        else
        {
            int noOfWasteCells = 0;
            for (int i = 0; i < k; i++)
            {
                if (wasteCells[i][0] < x || (wasteCells[i][0] == x && wasteCells[i][1] < y))
                {
                    noOfWasteCells++;
                }
            }
            int totalCells = (x - 1) * m + y;
            int totalCropCells = totalCells - noOfWasteCells;
            // cout << x << " " << y << " " << totalCells << " " << noOfWasteCells << " " << totalCropCells << " ";
            cout << crops[(totalCropCells - 1) % 3];
        }
        cout << "\n";
    }
    return 0;
}

/*
Renaissance-ProgrammingPathshala/M2/Mathematics/HomeWork_3/ColouredField.cpp
4 5 5 6
4 3
1 3
3 3
2 5
3 2
1 3
1 4
2 3
2 4
1 1
1 1

̧Input
4 5 5 6
4 3
1 3
3 3
2 5
3 2
1 3
1 4
2 3
2 4
1 1
1 1

Output
Waste
Grapes
Carrots
Kiwis
Carrots
Carrots
*/

File: Renaissance-ProgrammingPathshala/M2/Mathematics/HomeWork_3/ColouredField.py
# your code goes here
modulo = 3

crops = ['Carrots', 'Kiwis', 'Grapes']

n, m, k, t = map(int, input().split())

wasteland_data = [[] for i in range(n)]

for i in range(k):
    x, y = map(int, input().split())
    wasteland_data[x-1].append(y-1)

total_wastelands = 0
prefix_wastelands_sum = [0]*n
for i in range(1, n):
    total_wastelands += len(wasteland_data[i-1])
    prefix_wastelands_sum[i] = total_wastelands


for i in range(t):
    x, y = map(int, input().split())
    if y-1 in wasteland_data[x-1]:
        print('Waste')
    else:
        wate_land_count = 0
        for idx in wasteland_data[x-1]:
            if idx < y-1:
                wate_land_count += 1

        print(crops[((x-1)*m+y-1-wate_land_count -
              prefix_wastelands_sum[x-1]) % modulo])


"""
Renaissance-ProgrammingPathshala/M2/Mathematics/HomeWork_3/ColouredField.py
4 5 5 6  
4 3  
1 3  
3 3  
2 5  
3 2  
1 3  
1 4  
2 3  
2 4  
1 1  
1 1

Input
4 5 5 6  
4 3  
1 3  
3 3  
2 5  
3 2  
1 3  
1 4  
2 3  
2 4  
1 1  
1 1

Output
Waste  
Grapes  
Carrots  
Kiwis  
Carrots  
Carrots

"""
